### ISSUE TITLE
yolact onnx2ncnn的模型在windows上预测，得到的mask数据不对

### ISSUE THREAD

**Issue by junqiCao** [Tencent/ncnn#2248]
我是使用的pytorch1.4，在ubuntu16.04的系统上将yolact.onnx转化成yolact-sim.onnx，然后再windows上用onnx2ncnn转化成yolact.param和yolact.bin,然后又用ncnnoptimize优化了模型yolact-opt.param,yolact-opt.bin，然后微调了模型将3个0=-3改为0=0，然后在程序中yolact.cpp中调用，只有mask的结果是不对的，得到的图像数据很多负数，导致mask的结果不对。
模型参数如下：
Convolution              615                      1 1 614 616 0=32 1=1 5=1 6=8192 9=1
Permute                    617                      1 1 616 617 0=3
Concat                      813                      5 1 631 670 709 748 787 813 0=0
Concat                      814                      5 1 643 682 721 760 799 814 0=0
Concat                      815                      5 1 656 695 734 773 812 815 0=0
Softmax                    817                      1 1 814 817 0=1 1=1
程序中使用的是
	ex.extract("616", maskmaps); // 138x138 x 32
	ex.extract("813", location);   // 4 x 19248
	ex.extract("815", mask);       // maskdim 32 x 19248
	ex.extract("817", confidence); // 81 x 19248
大佬帮忙看下是哪里的问题，谢谢。

**Comment by junqiCao** [comment]
是不是所有的模型都能够成功转化，如果不能，那需要怎么定位问题呢？

**Comment by nihui** [comment]
是根据这文章的步骤操作下来的吗
https://zhuanlan.zhihu.com/p/128974102
另外，yolact ncnn 模型在 https://github.com/nihui/ncnn-assets/tree/master/models 也有，可以试试看

**Comment by junqiCao** [comment]
这篇文章看了好多遍了，用了下载的模型是没有问题的




    


在2020年10月23日 19:34，nihui 写道： 
是根据这文章的步骤操作下来的吗
https://zhuanlan.zhihu.com/p/128974102
另外，yolact ncnn 模型在 https://github.com/nihui/ncnn-assets/tree/master/models 也有，可以试试看

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/Tencent/ncnn/issues/2248#issuecomment-715287109",
"url": "https://github.com/Tencent/ncnn/issues/2248#issuecomment-715287109",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by junqiCao** [comment]
@nihui 您好，我今天又试了一下，发现在调用 https://github.com/nihui/ncnn-assets/tree/master/models 中的yolact.param 和yolact.bin的模型都能够正常显示，但是在使用自己的模型时，mask可以获取到128*128的图像数据，但是在cv::resize()这一步就报错，搞不懂为什么会在这个地方报错，检测框和mask、confidence也是对的，同样的程序，换一个模型就报莫名奇妙的错误。

**Comment by junqiCao** [comment]
失败的经历，给后人以警示：我在训练时使用的max_size为550，但是在转化模型时的使用的max_size为512，导致模型转化失败，得到一个经验教训：做事情要认认真真。

**Comment by nihui** [comment]
> 失败的经历，给后人以警示：我在训练时使用的max_size为550，但是在转化模型时的使用的max_size为512，导致模型转化失败，得到一个经验教训：做事情要认认真真。

感谢分享！

**Comment by junqiCao** [comment]
@Nihui ,我的图片是灰度图，如果想加快速度，有什么方法吗？




    


在2020年10月30日 13:00，nihui 写道： 

失败的经历，给后人以警示：我在训练时使用的max_size为550，但是在转化模型时的使用的max_size为512，导致模型转化失败，得到一个经验教训：做事情要认认真真。

感谢分享！

—You are receiving this because you modified the open/close state.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/Tencent/ncnn/issues/2248#issuecomment-719172809",
"url": "https://github.com/Tencent/ncnn/issues/2248#issuecomment-719172809",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by areebsyed** [comment]
> I use pytorch1.4. I converted yolact.onnx to yolact-sim.onnx on the ubuntu16.04 system, and then used onnx2ncnn to convert to yolact.param and yolact.bin on windows, and then optimized the model with ncnnoptimize yolact-opt.param, yolact-opt.bin, and then fine-tuned the model to change 3 0=-3 to 0=0, and then call it in yolact.cpp in the program. Only the result of the mask is wrong, the resulting image There are many negative numbers in the data, resulting in incorrect mask results.
> The model parameters are as follows:
> Convolution 615 1 1 614 616 0=32 1=1 5=1 6=8192 9=1
> Permute 617 1 1 616 617 0=3
> Concat 813 5 1 631 670 709 748 787 813 0=0
> Concat 814 5 643 682 721 760 799. 1 814 0 0 =
> Concat. 5. 1 815 656 695 734 773 812 815 0 = 0
> the Softmax 0 817 817 = 814. 1. 1. 1. 1. 1 =
> used in the program is
> ex.extract("616", maskmaps); // 138x138 x 32
> ex.extract("813", location); // 4 x 19248
> ex.extract("815", mask); // maskdim 32 x 19248
> ex .extract("817", confidence); // 81 x 19248
> big guys help to see where the problem is, thank you.

How do you fine tuned the model? which file and which line do you change?
@nihui

**Comment by chingi071** [comment]
@nihui @junqiCao 您好，我想请问一下您将yolact.onnx转化成yolact-sim.onnx时，有出现这个错误吗?
我执行 $ python eval.py --trained_model=weights/yolact_resnet50_54_800000.pth --score_threshold=0.15 --top_k=15 --image=test.jpg 指令会出现以下错误，不过可以输出 yolact-sim.onnx
但在windows 或 linux上用 ncnnoptimize优化模型yolact-opt.param,yolact-opt.bin时会出现 Segmentation fault (core dumped) 
想请教一下您们是怎么处理的，谢谢。
```
TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if self.last_img_size != (cfg._tmp_img_w, cfg._tmp_img_h):
```

**Comment by huyunlei** [comment]
@chingi071   请问您这个问题解决了吗， eval.py 问你题，我也遇到了 ，愁死了，百度了一下午 ，查不出来，您要是解决了能不能告诉我一下，万分感谢， 我先磕一个-.-
![image](https://user-images.githubusercontent.com/57059809/113862579-29cf1000-97db-11eb-9a4d-b77abc198ae9.png)
![image](https://user-images.githubusercontent.com/57059809/113862581-29cf1000-97db-11eb-8c01-18c4611eb8d8.png)
![image](https://user-images.githubusercontent.com/57059809/113862580-29cf1000-97db-11eb-875b-ad9ea30ba7a1.png)

**Comment by huyunlei** [comment]
@chingi071 美女小姐姐，你这个问题解决了吗 ，

**Comment by chingi071** [comment]
@huyunlei 您好，這個問題我也還沒解決....

**Comment by huyunlei** [comment]
@chingi071  好的，谢谢

**Comment by huyunlei** [comment]
@chingi071  还是  很感谢你 回复的 -

**Comment by chingi071** [comment]
@huyunlei 不會~

**Comment by Yaifan** [comment]
大神们解决了吗？

**Comment by huyunlei** [comment]
这是来自QQ邮箱的自动回复邮件。
 
您好，您发来的邮件已经收到，我会尽快给您回复。

**Comment by 74284853** [comment]
> @nihui ,我的图片是灰度图，如果想加快速度，有什么方法吗？ 在2020年10月30日 13:00，nihui 写道： 失败的经历，给后人以警示：我在训练时使用的max_size为550，但是在转化模型时的使用的max_size为512，导致模型转化失败，得到一个经验教训：做事情要认认真真。 感谢分享！ —You are receiving this because you modified the open/close state.Reply to this email directly, view it on GitHub, or unsubscribe. [ { "@context": "http://schema.org", "@type": "EmailMessage", "potentialAction": { "@type": "ViewAction", "target": "[#2248 (comment)](https://github.com/Tencent/ncnn/issues/2248#issuecomment-719172809)", "url": "[#2248 (comment)](https://github.com/Tencent/ncnn/issues/2248#issuecomment-719172809)", "name": "View Issue" }, "description": "View this Issue on GitHub", "publisher": { "@type": "Organization", "name": "GitHub", "url": "https://github.com" } } ]

我遇到了同样的错误，请问’导出时max_size为512的错误‘指的是修改代码哪里呢？希望可以得到您的回复

**Comment by huyunlei** [comment]
这是来自QQ邮箱的自动回复邮件。
 
您好，您发来的邮件已经收到，我会尽快给您回复。

**Comment by 74284853** [comment]
请问您有注意到这条消息吗？期待您的回复！！



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年3月13日(周一) 上午10:01
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [Tencent/ncnn] yolact onnx2ncnn的模型在windows上预测，得到的mask数据不对 (#2248)




 这是来自QQ邮箱的自动回复邮件。
 &nbsp;
 您好，您发来的邮件已经收到，我会尽快给您回复。

**Comment by 74284853** [comment]
> 失败的经历，给后人以警示：我在训练时使用的max_size为550，但是在转化模型时的使用的max_size为512，导致模型转化失败，得到一个经验教训：做事情要认认真真。

在復現過程中出現了相同的mask預測錯誤，請問max_size为512是在哪裏設置的呢？

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx
